### PR TITLE
Fix multi nccl comm & wait server ready

### PR DESCRIPTION
**Pull request by wangxicoding**
### PR types

Others
### PR changes

Others
### Describe

- Fix fleet nccl_comm_num always=1. 
- When sync_nccl_allreduce, set num_threads = max(nccl_comm_num + 1, num_threads) instead of nccl_comm_num + 1.
- When nccl_comm_num > 1, if sync_nccl_allreduce=True, nccl comms will not overlap. Add warning, the user should set sync_nccl_allreduce=False to make sure different nccl comms can overlap
image
- Fix graph_execution_optimizer not wait server ready, GRPC may be failed.

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_optimizers/graph_execution_optimizer.py [hunk 1]
```diff
@@ -148,13 +148,12 @@ def _try_to_compile(self, startup_program, main_program, loss):
...
-                exe_strategy.num_threads = 2 * local_build_strategy.nccl_comm_num + 1
-            if exe_strategy.num_threads > 4:
+            exe_strategy.num_threads = max(
+                local_build_strategy.nccl_comm_num + 1,
+                exe_strategy.num_threads)
+            if local_build_strategy.nccl_comm_num > 1:
                 logging.warn(
-                    "if you use hierachical_allreduce or "
-                    "with multi nccl comm, please set distributed_strategy.sync_nccl_allreduce=False"
+                    "nccl_comm_num > 1, you may need to set sync_nccl_allreduce=False to ensure that different nccl comms can overlap"
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_optimizers/graph_execution_optimizer.py:164]: 用户如果设置了还会显示warning是吗？

Comment (wangxicoding) [reply python/paddle/distributed/fleet/meta_optimizers/graph_execution_optimizer.py:164]: 这个是在sync_allreduce里面，如果设置了sync_nccl_allreduce=False就不会显示了

File: python/paddle/distributed/fleet/meta_optimizers/graph_execution_optimizer.py [hunk 2]
```diff
@@ -51,13 +52,21 @@ def backward(self,
     # should fix the variable
     def _setup_nccl_op(self, startup_program, main_program, build_strategy):
         trainer_endpoints = self.role_maker._get_trainer_endpoints()
-        trainers = trainer_endpoints
+        other_trainers = copy.copy(trainer_endpoints)
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_optimizers/graph_execution_optimizer.py:55]: 这段代码似曾相识呀！

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by chalsliu — APPROVED** [review]

**Pull request merged**